msprime/demography.py

```diff
@@ -3459,28 +3459,16 @@ def _proportions_to_sequential(P):
     return C
 
 
-def _matrix_exponential_eigen(A):
+def _matrix_exponential(A, n=5):
     """
     Returns the matrix exponential of A.
-    https://en.wikipedia.org/wiki/Matrix_exponential
-    Note: this is not a general purpose method and is only intended for use within
-    msprime.
-    """
-    d, Y = np.linalg.eig(A)
-    Yinv = np.linalg.pinv(Y)
-    D = np.diag(np.exp(d))
-    B = np.matmul(Y, np.matmul(D, Yinv))
-    return np.real_if_close(B, tol=1000)
-
-
-def _matrix_exponential_poisson(A, n=5):
-    """
     Only works if the offdiagonals of A are nonnegative
     and the row sums of A are less than or equal to zero.
+    Melloy & Bennett (1993), https://doi.org/10.1016/0377-0427(93)90036-B
     """
-    # TODO: check if these affect the timing
     assert np.max(np.diag(A)) <= 0
-    assert np.min(A - np.diag(np.diag(A))) >= 0
+    assert np.min(np.tril(A, k=-1)) == 0
+    assert np.min(np.triu(A, k=1)) == 0
```

Comment:
oh good! I was hoping there was a better way to do that.

Reply:
I did some informal timing tests, and was surprised that I could measure a difference with/without the asserts. This change made the time differences harder to measure (closer to the noise).